nixos/modules/system/boot/kernel.nix

```diff
@@ -54,9 +54,76 @@ in
     };
 
     boot.kernelParams = mkOption {
+      type = let
+        coerce = list: let
+          mkItem = item: let
+            splitted = builtins.match "([^=]*)=(.*)" item;
+          in if splitted == null then nameValuePair item true
+             else nameValuePair (head splitted) (last splitted);
+        in listToAttrs (map mkItem list);
+        targetType = with types; attrsOf (either (either bool int) str);
+      in with types; coercedTo (listOf str) coerce targetType;
+      default = {};
+      example = {
+        acpi_apic_instance = 2;
+        tmem = true;
+        nf_conntrack.acct = 1;
+        acpi_osi = "Windows 2000";
+      };
+      description = let
+        docUrl = "https://www.kernel.org/doc/html/latest/admin-guide/"
+               + "kernel-parameters.html";
+      in ''
+        An attribute set of kernel command line paramaters, where the attribute
+        name denotes a particular parameter. If the attribute value is
+        <literal>true</literal> the attribute name is passed verbatim and if
+        it's <literal>false</literal> the parameter is dropped entirely.
+
+        Have a look at the <link xlink:href="${docUrl}">documentation</link>
+        for possible values.
+      '';
+    };
+
+    boot.initArgs = mkOption {
+      type = types.listOf types.str;
+      default = [];
+      example = [ "--crash-shell" "--crash-vt=1" ];
+      description = ''
+        Arguments passed to <citerefentry>
+          <refentrytitle>init</refentrytitle>
+          <manvolnum>1</manvolnum>
+        </citerefentry> via kernel parameters.
+
+        You could directly use <option>boot.kernelParams</option> to do the
+        same thing, but if there are collisions with existing kernel parameters
+        it's very likely that the init argument is silently ignored.
+
+        Furthermore, passing init arguments via
+        <option>boot.kernelParams</option> will cause the arguments to be
+        sorted, so it's unsuitable if you use an init that depends on the order
+        of arguments.
```

Comment:
Are there any init arguments that depend on the order? `systemd(1)` suggests the order shouldn't matter.

Reply:
There aren't, but init could also be a shell script and/or basically anything, if you want to pass `$1` to such a shell script it will be difficult if the value is sorted.

Reply:
Well, in NixOS init is always systemd.

Reply:
By default it is, but for example to build for embedded systems, it might make sense to change it. An example of this is in the [VirtualBox tests](https://github.com/NixOS/nixpkgs/blob/49b17a81270535179458050ff648064edf87be0f/nixos/tests/virtualbox.nix#L40-L44) which changes init to avoid running a full system during software virtualization (which already is slow as hell).